internal/hashing/argon2id.go

```diff
@@ -0,0 +1,162 @@
+var DefaultArgon2IDParameters = Argon2IDParameters{
+	SaltLength: 16,
+	Time:       2,
+	Memory:     19 * 1024, // 19 MiB
+	Threads:    1,
+	KeyLength:  32,
+}
+
+var ErrInvalidPassword = errors.New("invalid password")
+
+// Takes the password string, makes sure it's not empty, and hashes it using argon2id
+func HashPassword(password string) (string, error) {
+	if strings.TrimSpace(password) == "" {
+		return "", fmt.Errorf("%w: password cannot be empty", ErrInvalidPassword)
+	}
+	salt, err := generateSalt(DefaultArgon2IDParameters.SaltLength)
+	if err != nil {
+		return "", err
+	}
+	hash := hashPasswordWithSaltAndParams(password, salt, DefaultArgon2IDParameters)
+
+	encoded_salt := base64.RawStdEncoding.EncodeToString(salt)
+	encoded_hash := base64.RawStdEncoding.EncodeToString(hash)
+	hashedPassword := fmt.Sprintf("$argon2id$v=%d$m=%d,t=%d,p=%d$%s$%s",
+		ARGON2_VERSION,
+		DefaultArgon2IDParameters.Memory,
+		DefaultArgon2IDParameters.Time,
+		DefaultArgon2IDParameters.Threads,
+		encoded_salt,
+		encoded_hash,
+	)
+	return string(hashedPassword), nil
+}
+
+// CompareHashAndPassword takes a hashed password string and a password,
+// hashes the password with the same parameters as the hashed password and
+// compares the resulting value, returning an error if they do not match.
+//
+// When this function is passed an invalid hashed password string, it will
+// hash the password with the default parameters to prevent an attacker from
+// getting information from the timing of a login failure.
+func CompareHashAndPassword(hashedPassword string, password string) error {
+	params, salt, hash, err := parseArgon2IDPHC(hashedPassword)
+	if err != nil {
+		// The hashedPassword was not parseable, hash the password with default values
+		// to spend the same amount of time.
+		_, _ = HashPassword(password)
+		return fmt.Errorf("password did not match")
+	}
+	newHash := hashPasswordWithSaltAndParams(password, salt, params)
+	if string(newHash) == string(hash) {
+		return nil
+	}
+	return fmt.Errorf("password did not match")
+}
+
+func hashPasswordWithSaltAndParams(password string, salt []byte, params Argon2IDParameters) []byte {
+	return argon2.IDKey(
+		[]byte(password),
+		salt,
+		params.Time,
+		params.Memory,
+		params.Threads,
+		params.KeyLength,
+	)
+}
+
+func generateSalt(n uint32) ([]byte, error) {
+	b := make([]byte, n)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+// Parses a PHC string for Argon2id, and returns Argon2id parameters, the salt and the hash
+func parseArgon2IDPHC(phc string) (Argon2IDParameters, []byte, []byte, error) {
+	fields := strings.Split(phc, "$")
+	if len(fields) != 6 {
+		return DefaultArgon2IDParameters, nil, nil, fmt.Errorf("cannot parse hashed password strings for argon2id")
+	}
+	if fields[1] != "argon2id" || fields[2] != fmt.Sprintf("v=%d", ARGON2_VERSION) {
+		return DefaultArgon2IDParameters, nil, nil, fmt.Errorf("cannot parse hashed password strings for argon2id")
+	}
```

Comment:
I would encourage to separate these into 2 conditions and have more specific error messages (ex. `mode is not argondID: <whatever mode was found>`, `version not 19: <whatever version was found>`.

Reply:
I went for a single error message here because this is a package scoped function that is only used in a single place and we do not do anything with the error. I used simple `fmt.Errorf` as those are unrecoverable errors, and followed PL007 here.

I am not against making the different errors more explicit, but I doubt there is value here, as they should not be logged (this code will get called on every login, including with invalid users, so an attacker could fill out the logs with a DoS attack), and the outcome is the same in all cases; we cannot decode the hash, so the password will not match.

Reply:
I changed to specific errors.